### PR TITLE
Add support for manually supplying Bedrock resource packs

### PR DESCRIPTION
**Pull request by rtm516**
This adds support for loading bedrock resource packs in and sending them to clients.
You can use https://ozelot379.github.io/ConvertJavaTextureToBedrock/ to convert Java resource packs.

This does not automatically convert or send them from the Java server, it only send already known resource packs to the client.

### CODE REVIEW

File: connector/src/main/java/org/geysermc/connector/network/UpstreamPacketHandler.java [hunk 1]
```diff
@@ -73,13 +86,44 @@ public boolean handle(ResourcePackClientResponsePacket packet) {
...
+                for(String id : packet.getPackIds()) {
+                    ResourcePackDataInfoPacket data = new ResourcePackDataInfoPacket();
+                    String[] packID = id.split("_");
+                    ResourcePack pack = ResourcePack.PACKS.get(packID[0]);
+                    ResourcePackManifest.Header header = pack.getManifest().getHeader();
+
+                    data.setPackId(header.getUuid());
+                    int chunkCount = (int) Math.ceil((int) pack.getFile().length() / (double) ResourcePack.CHUNK_SIZE);
+                    data.setChunkCount(chunkCount);
+                    //data.setChunkCount(pack.getFile().length()/ResourcePack.CHUNK_SIZE);
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/UpstreamPacketHandler.java:100]: We can remove this.

File: connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
+package org.geysermc.connector.utils;
+
+import org.geysermc.connector.GeyserConnector;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.ZipFile;
+
+public class ResourcePack {
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java:10]: I would add a small Javadoc comment here explaining the purpose of this class, e.g.
```java
/**
* Represents a resource pack in the `packs` folder.
*/
```

File: connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    public static final Map<String, ResourcePack> PACKS = new HashMap<>();
+    public static final int CHUNK_SIZE = 102400;
+
+    private byte[] sha256;
+    private File file;
+    private ResourcePackManifest manifest;
+    private ResourcePackManifest.Version version;
+
+    public static void loadPacks() {
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java:19]: I would add a Javadoc comment here.

File: connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                                pack.version = ResourcePackManifest.Version.fromArray(manifest.getHeader().getVersion());
+
+                                PACKS.put(pack.getManifest().getHeader().getUuid().toString(), pack);
+                            } catch (Exception e) {
+                                e.printStackTrace();
+                            }
+                        }
+                    });
+                } catch (Exception e) {
+                    GeyserConnector.getInstance().getLogger().error(file.getName() + " " + "is broken!");
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java:51]: This should be a translation string with maybe a little more detail explaining that it is a file that is broken.

File: connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import org.geysermc.connector.GeyserConnector;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.ZipFile;
+
+public class ResourcePack {
+    public static final Map<String, ResourcePack> PACKS = new HashMap<>();
+    public static final int CHUNK_SIZE = 102400;
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java:12]: Javadoc comment here.

File: connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.geysermc.connector.GeyserConnector;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.ZipFile;
+
+public class ResourcePack {
+    public static final Map<String, ResourcePack> PACKS = new HashMap<>();
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java:11]: Javadoc comment here.

File: connector/src/main/java/org/geysermc/connector/network/UpstreamPacketHandler.java [hunk 7]
```diff
@@ -73,13 +86,44 @@ public boolean handle(ResourcePackClientResponsePacket packet) {
...
+
             case HAVE_ALL_PACKS:
-                ResourcePackStackPacket stack = new ResourcePackStackPacket();
-                stack.setExperimental(false);
-                stack.setForcedToAccept(false);
-                stack.setGameVersion("*");
-                session.sendUpstreamPacket(stack);
+                ResourcePackStackPacket stackPacket = new ResourcePackStackPacket();
+                stackPacket.setExperimental(false);
+                stackPacket.setForcedToAccept(true);
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/UpstreamPacketHandler.java:115]: Should we have a setting to toggle this?

File: connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    private byte[] sha256;
+    private File file;
+    private ResourcePackManifest manifest;
+    private ResourcePackManifest.Version version;
+
+    /**
+     * Loop through the packs directory and locate an valid resource pack files
+     */
+    public static void loadPacks() {
+        File directory = new File("packs");
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/utils/ResourcePack.java:33]: This should be relative to the config directory.

File: connector/src/main/java/org/geysermc/connector/network/session/GeyserSession.java [hunk 9]
```diff
@@ -554,7 +554,7 @@ private void startGame() {
         startGamePacket.setItemEntries(ItemRegistry.ITEMS);
         startGamePacket.setVanillaVersion("*");
         // startGamePacket.setMovementServerAuthoritative(true);
-        upstream.sendPacket(startGamePacket);
+        upstream.sendPacketImmediately(startGamePacket);
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/session/GeyserSession.java:557]: Why this change?

Comment (rtm516) [reply connector/src/main/java/org/geysermc/connector/network/session/GeyserSession.java:557]: Not sure, it can be reverted.

**Comment by RednedEpic** [comment]
Needs rebasing and merge conflicts to be resolved.

File: connector/src/main/java/org/geysermc/connector/utils/FileUtils.java [hunk 10]
```diff
@@ -50,6 +55,16 @@
         return objectMapper.readValue(src, valueType);
     }
 
+    public static <T> T loadYaml(InputStream src, Class<T> valueType) throws IOException {
+        ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory()).enable(JsonParser.Feature.IGNORE_UNDEFINED).disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+        return objectMapper.readValue(src, valueType);
+    }
+
+    public static <T> T loadJson(InputStream src, Class<T> valueType) throws IOException {
+        ObjectMapper objectMapper = new ObjectMapper(new JsonFactory()).enable(JsonParser.Feature.IGNORE_UNDEFINED).enable(JsonParser.Feature.ALLOW_COMMENTS).disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/utils/FileUtils.java:64]: `GeyserConnector` defines a static OBJECT_MAPPER instance; is that possible to be used here?

**Comment by LoneDev6** [comment]
Would be cool to get automatic CustomModelData to namespacedid conversion, I have no knowledge on how Bedrock managed that custom items stuff, I'd love to help

**Review by DoctorMacc — APPROVED** [review]

**Comment by d1ddle** [comment]
Could someone make a guide/step by step usage of this if it is even usable yet?
thanks :P

**Comment by AnhvietPlays** [comment]
I would love to see ItemsAdder guide on how to do this. Thank You!

**Comment by AnhvietPlays** [comment]
> > I would love to see ItemsAdder guide on how to do this. Thank You!
> 
> me too

me as well
